homeassistant/components/twitch/sensor.py

```diff
@@ -130,6 +131,10 @@ def __init__(self, channel: TwitchUser, client: Twitch) -> None:
 
     async def async_update(self) -> None:
         """Update device state."""
+        # First, check if the token is still valid
+        if await self._client.get_refreshed_user_auth_token() is None:
```

Comment:
So currently, home assistant is responsible for getting the token. twitchAPI also has capabilities to request/refresh/manage the token.

Theoretically we could use twitch API for this, but I think I'm leaning towards having HA mane this, so we don't have to write code to update the token in the config entry data. 

This means we have to also pass in the oauth2 session and check if the token is valid. It would automatically request a new one (and save it if not).

The downside of having this code in the async_update function is that there is a change we are going to do every channel update at the same time, thus also request tokens again for every channel you have set up. (This could be solved for now by setting a PARALLEL_UPDATES variable in the sensor.py and setting it to 1 so only 1 sensor updates at a time). Later this could be replaced with a coordinator.

Reply:
Setting this variable may also help with twitch rate limits when requesting data for many many channels. I think we need a coordinator, which I made an attempt at in the past, but it ended up getting abandoned for some reason or another.

Would be nice if twitch had a nicer api for requesting data similar to github's graphql api

Reply:
Doing this does slow the startup down to what I'd call glacial. Might need to jump straight to a coordinator

Reply:
Yeah running with multiple followed channels causes this to be timed out. Closing